Binding library internal delegate generated as public

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨 -->

<!--
Please fill in the *entire* template below.
The template is designed to avoid unnecessary delays to confirm and fix reported issues. Issues with missing information generally end up with a `need-info` label and take significantly more time to fix. Please help us help you!
For more information see https://github.com/xamarin/xamarin-macios/wiki/Submitting-Bugs-&-Suggestions
For support requests use https://www.xamarin.com/support
-->
### Steps to Reproduce

1. Create an iOS binding library project on .NET 6
2. Add a delegate in the ApiDefinition.cs: 
```
    [Internal]
    internal delegate SentryBreadcrumb SentryBeforeBreadcrumbCallback(SentryBreadcrumb arg0);
```
3. Type in argument/return of the delegate also has `[Internal]` and is generated with `internal` access modifier.
4. `dotnet build`

Repro below.

### Expected Behavior

Builds

### Actual Behavior

```
/Users/bruno/git/sentry-dotnet/src/Sentry/obj/Debug/net6.0-ios/iOS/SupportDelegates.g.cs(52,46): error CS0058: Inconsistent accessibility: return type 'SentryBreadcrumb' is less accessible than delegate 'SentryBeforeBreadcrumbCallback' [/Users/bruno/git/sentry-dotnet/src/Sentry/Sentry.csproj]
```

```
/Users/bruno/git/sentry-dotnet/src/Sentry/obj/Debug/net6.0-ios/iOS/SupportDelegates.g.cs(52,46): error CS0058: Inconsistent accessibility: return type 'SentryBreadcrumb' is less accessible than delegate 'SentryBeforeBreadcrumbCallback' [/Users/bruno/git/sentry-dotnet/src/Sentry/Sentry.csproj]
/Users/bruno/git/sentry-dotnet/src/Sentry/obj/Debug/net6.0-ios/iOS/SupportDelegates.g.cs(52,46): error CS0059: Inconsistent accessibility: parameter type 'SentryBreadcrumb' is less accessible than delegate 'SentryBeforeBreadcrumbCallback' [/Users/bruno/git/sentry-dotnet/src/Sentry/Sentry.csproj]
```

Generated code:

```
namespace Sentry.iOS {
	#nullable enable
	public delegate Sentry.iOS.SentryBreadcrumb SentryBeforeBreadcrumbCallback (Sentry.iOS.SentryBreadcrumb arg0);
	public delegate Sentry.iOS.SentryEvent SentryBeforeSendEventCallback (Sentry.iOS.SentryEvent arg0);
}
```

### Environment

<!--
1. How to get version information: https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#version-information
2. Paste into the code block below (between ```)
-->

<details>
<summary>Version information</summary>

workloads:

```
Installed Workload Ids      Installation Source
-----------------------------------------------
maui-ios                    SDK 6.0.300
maui-android                SDK 6.0.300
ios                         SDK 6.0.300
maui                        SDK 6.0.300
android                     SDK 6.0.300

```
SDK/runtime:

```
.NET SDK (reflecting any global.json):
 Version:   6.0.301
 Commit:    43f9b18481

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  12.4
 OS Platform: Darwin
 RID:         osx.12-x64
 Base Path:   /usr/local/share/dotnet/sdk/6.0.301/

Host (useful for support):
  Version: 6.0.6
  Commit:  7cca709db2

.NET SDKs installed:
  2.1.818 [/usr/local/share/dotnet/sdk]
  3.1.414 [/usr/local/share/dotnet/sdk]
  3.1.415 [/usr/local/share/dotnet/sdk]
  3.1.416 [/usr/local/share/dotnet/sdk]
  3.1.419 [/usr/local/share/dotnet/sdk]
  3.1.420 [/usr/local/share/dotnet/sdk]
  5.0.402 [/usr/local/share/dotnet/sdk]
  5.0.403 [/usr/local/share/dotnet/sdk]
  5.0.404 [/usr/local/share/dotnet/sdk]
  5.0.408 [/usr/local/share/dotnet/sdk]
  6.0.100-rc.2.21505.57 [/usr/local/share/dotnet/sdk]
  6.0.100 [/usr/local/share/dotnet/sdk]
  6.0.101 [/usr/local/share/dotnet/sdk]
  6.0.106 [/usr/local/share/dotnet/sdk]
  6.0.202 [/usr/local/share/dotnet/sdk]
  6.0.300 [/usr/local/share/dotnet/sdk]
  6.0.301 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.20 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.22 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.25 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.26 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.11 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.13 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.17 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-rc.2.21480.10 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.4 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.20 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.22 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.25 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.26 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.11 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.13 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.17 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-rc.2.21480.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.1 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.4 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.5 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.6 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]
```

</details>

### Build Logs

<!--
1. How to get a build log: https://github.com/xamarin/xamarin-macios/wiki/Diagnosis#build-logs
2. Place the cursor below this comment block.
3. Drag and drop the build log.
-->

### Example Project (If Possible)

<!--
1. Place cursor below this comment block.
2. Drag and drop the compressed project or files needed to reproduce.
-->

Repro in this commit: [getsentry/sentry-dotnet@`be0f2e2` (#1727)](https://github.com/getsentry/sentry-dotnet/pull/1727/commits/be0f2e2aa9cfc7af1d15f77ed32f4c7210f4b701)
Part of this PR: https://github.com/getsentry/sentry-dotnet/pull/1727

<!--
Switch to the "Preview" tab to ensure your issue renders correctly.
-->

Actually, while it looks like `[NullAllowed]` is working on the input parameters, `[return: NullAllowed]` does not work.  And if I try to replace to workaround, I get `error CS8600: Converting null literal or possible null value to non-nullable type.` in `ObjCRuntime/Trampolines.g.cs`.

So there's two issues here.  Neither `[Internal]` nor `[return: NullAllowed]` work on delegate types.

Should I open a new issue for the latter? Or ok to track them both here?

--- a/src/bgen/Generator.cs
+++ b/src/bgen/Generator.cs
@@ -4799,10 +4799,12 @@ public partial class Generator : IMemberGatherer {
 				if (AttributeManager.HasAttribute (mi.DeclaringType, "MonoNativeFunctionWrapper"))
 					print ("[MonoNativeFunctionWrapper]\n");
 
-				print ("public delegate {0} {1} ({2});",
+				var accessibility = mi.DeclaringType.IsInternal (this) ? "internal" : "public";
+				print ("{3} delegate {0} {1} ({2});",
 					   RenderType (mi.ReturnType, mi.ReturnTypeCustomAttributes),
 					   shortName,
-					   RenderParameterDecl (mi.GetParameters ()));
+					   RenderParameterDecl (mi.GetParameters ()),
+					   accessibility);
 			}
 
 			if (group.Namespace is not null) {

--- a/tests/generator/BGenTests.cs
+++ b/tests/generator/BGenTests.cs
@@ -1420,6 +1420,12 @@ namespace GeneratorTests {
 			bgen.AssertExecute ("build");
 		}
 
+		[Test]
+		public void InternalDelegate ()
+		{
+			BuildFile (Profile.iOS, "tests/internal-delegate.cs");
+		}
+
 		BGenTool BuildFile (Profile profile, params string [] filenames)
 		{
 			return BuildFile (profile, true, false, filenames);

--- a/tests/generator/tests/internal-delegate.cs
+++ b/tests/generator/tests/internal-delegate.cs
@@ -1,0 +1,15 @@
+using System;
+using Foundation;
+
+namespace NS {
+	[Internal]
+	delegate void Whatever (SomeType value);
+
+	[Internal]
+	[BaseType (typeof (NSObject))]
+	interface SomeType {
+		[Export ("whatever")]
+		Whatever Nope { get; set; }
+	}
+}
+

